### PR TITLE
swap: prevent debt cheques

### LINKED ISSUE #1931: Prevent another node from tricking us into sending a cheque after receiving a cheque
Right now a node can trigger one of its peers to send a cheque simply by sending a cheque and thereby putting that peer into debt. We should probably prevent this as this might be used for attacks. (e.g. sending a cheque that would bounce to get a real cheque to steal funds).

### PR DESCRIPTION
**Pull request by mortelli**
This PR:
- extends the `processAndVerifyCheque` function so that it returns an error when receiving a cheque that would put the node into debt.
  - to determine this, the tentative balance (after processing the cheque) is calculated, and if lower than `0 - ChequeDebtTolerance`, an error is returned.
  - `ChequeDebtTolerance` is a new `const` set—for now—to roughly 20% of the payment threshold.
  - in principle, we should not accept any cheques that would put us into debt, but due to possible out-of-sync balances, we are allowing _some_ negative resulting balances (as well as any positive ones). 
  - this is tested by the new `TestDebtCheques` function.
- adds missing error handling in SWAP tests.
- removes `testMsgBigPrice` and `TestPingPongChequeSimulation`, as this test was considered out of place in the normal testing suite due to its stress-test nature.

closes #1931

### CODE REVIEW

**Comment by mortelli** [comment]
@holisticode i've observed failures when running something as:
```bash
/usr/local/go/bin/go test -count=3 -timeout 30s github.com/ethersphere/swarm/swap -run ^TestPingPongChequeSimulation$
```
but they only seem to happen as long as `count` is higher than `1`. 

i know there may be no real motivation to test this, but can you think of a reason why running `TestPingPongChequeSimulation` many times in parallel would cause a problem?

the error is related to the PR, since it looks like this:
```bash
rafa@rafa-laptop:~/go/src/github.com/ethersphere/swarm$ /usr/local/go/bin/go test -count=3 -timeout 30s github.com/ethersphere/swarm/swap -run ^TestPingPongChequeSimulation$
ERROR[01-13|14:58:30.562|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=08b7ca5afcfd9150 reason="hive stopping"
ERROR[01-13|14:58:31.000|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=ab402a0b39f3057c reason="message handler: (msg code 1): processing and verifying received cheque: received cheque would result in balance -19633281260474 which exceeds tolerance 1963328126047 and would cause debt"
ERROR[01-13|14:58:31.000|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=ab402a0b39f3057c reason="peer ab402a0b39f3057c2f14b869fb9d333300ac30800c3ed9ddedb8295b7a6f0abf not a swap enabled peer"
ERROR[01-13|14:58:51.002|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=c1b7b2bc8fda37c4 reason="hive stopping"
--- FAIL: TestPingPongChequeSimulation (20.44s)
    simulations_test.go:345: one or more wait routines timed out: Timed out waiting for cheque to have been cached - Timed out waiting for cheque to be confirmed
ERROR[01-13|14:58:51.775|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=b16fa940572c0942 reason="hive stopping"
FAIL
FAIL	github.com/ethersphere/swarm/swap	22.005s
FAIL
```

**Comment by mortelli** [comment]
> @holisticode i've observed failures when running something as:
> 
> ```shell
> /usr/local/go/bin/go test -count=3 -timeout 30s github.com/ethersphere/swarm/swap -run ^TestPingPongChequeSimulation$
> ```
> 
> but they only seem to happen as long as `count` is higher than `1`.
> 
> i know there may be no real motivation to test this, but can you think of a reason why running `TestPingPongChequeSimulation` many times in parallel would cause a problem?
> 
> the error is related to the PR, since it looks like this:
> 
> ```shell
> rafa@rafa-laptop:~/go/src/github.com/ethersphere/swarm$ /usr/local/go/bin/go test -count=3 -timeout 30s github.com/ethersphere/swarm/swap -run ^TestPingPongChequeSimulation$
> ERROR[01-13|14:58:30.562|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=08b7ca5afcfd9150 reason="hive stopping"
> ERROR[01-13|14:58:31.000|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=ab402a0b39f3057c reason="message handler: (msg code 1): processing and verifying received cheque: received cheque would result in balance -19633281260474 which exceeds tolerance 1963328126047 and would cause debt"
> ERROR[01-13|14:58:31.000|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=ab402a0b39f3057c reason="peer ab402a0b39f3057c2f14b869fb9d333300ac30800c3ed9ddedb8295b7a6f0abf not a swap enabled peer"
> ERROR[01-13|14:58:51.002|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=c1b7b2bc8fda37c4 reason="hive stopping"
> --- FAIL: TestPingPongChequeSimulation (20.44s)
>     simulations_test.go:345: one or more wait routines timed out: Timed out waiting for cheque to have been cached - Timed out waiting for cheque to be confirmed
> ERROR[01-13|14:58:51.775|github.com/ethersphere/swarm/p2p/protocols/protocol.go:248] dropping peer with DiscSubprotocolError  peer=b16fa940572c0942 reason="hive stopping"
> FAIL
> FAIL	github.com/ethersphere/swarm/swap	22.005s
> FAIL
> ```

forgot [we had already ran into this before](https://github.com/ethersphere/swarm/issues/1931#issuecomment-561780135)

**Comment by holisticode** [comment]
Strictly speaking that is not the same, although the effect looks like the same.

When we first had this problem, we identified that it was due to how the protocol was handling sending and receiving: it would first do the accounting and only then send the message, which would have the effect that if a threshold crossing happened, then it would issue a Cheque. This gave reason to Pr https://github.com/ethersphere/swarm/pull/2051. 

So this is now addressed, and if messages arrive in a different order, **it has a different reason**, and it is not the same issue.

**Comment by mortelli** [comment]
i have repeated the (now modified) `TestMultiChequeSimulation` test to determine how often cheques are wrongly perceived as debt cheques due to out-of-order messages.

the test conditions are:
- the message price is set to 1/100 of the value of the payment threshold.
- the cheque debt tolerance is set to 1/5 of the value of the payment threshold.

out of 1000 times, i got 4 failures due to cheques being perceived as debt-inducing. 

i am assuming this is caused by out of order protocol messages, but i don't know for sure. it would be nice to verify this with better logging, or some other tool. 

if the message price was set to a smaller fraction of the payment threshold, this would probably have an impact on this result as well.

File: swap/simulations_test.go [hunk 1]
```diff
@@ -458,27 +330,47 @@ func TestMultiChequeSimulation(t *testing.T) {
...
+			if err := waitForChequeProcessed(t, params.backend, counter, lastCount, debitorSvc.swap.peers[creditor], expectedPayout); err != nil {
+				t.Fatal(err)
+			}
+			expectedPayout += uint64(-balanceAfterMessage)
+		}
+
+		lastCount++
 	}
+	// give enough time for all messages to be processed
+	time.Sleep(10 * time.Millisecond)
```
Comment (ralph-pichler) [inline swap/simulations_test.go]: maybe it would be better to wait here by watching some metric (like in `TestBasicSwapSimulation`) instead of an arbitrary timeout?

File: swap/swap_test.go [hunk 2]
```diff
@@ -682,6 +688,125 @@ func TestResetBalance(t *testing.T) {
...
+	// balance should now be -ChequeDebtTolerance according to the debitor
+	if debitor.getBalance() != -int64(ChequeDebtTolerance) {
+		t.Fatalf("expected balance to be %d for peer %v, but it is %d", -int64(ChequeDebtTolerance), debitor.ID(), debitor.getBalance())
+	}
+	// no cheque should be pending
+	if creditor.getPendingCheque() != nil {
+		t.Fatalf("expected no pending cheque for peer %v, but found %v", creditor.ID(), creditor.getPendingCheque())
+	}
+	// verify last received cheque
+	if debitor.getLastReceivedCheque() != cheque {
```
Comment (ralph-pichler) [inline swap/swap_test.go]: use `.Equal` here

File: swap/swap_test.go [hunk 3]
```diff
@@ -682,6 +688,125 @@ func TestResetBalance(t *testing.T) {
...
+	// cheque should not have gone through as it would put the creditor in debt
+	if err == nil {
+		t.Fatal("expected invalid cheque reception to cause a failure, but got none")
+	}
+	// since the emission caused an error, the confirmation should not be simulated
+	// cheque should not be marked as sent, and should be pending
+	if creditor.getLastSentCheque() != nil {
+		t.Fatalf("expected no cheque sent to peer %v, but it is %v", creditor.ID(), creditor.getLastSentCheque())
+	}
+	if creditor.getPendingCheque() != cheque {
```
Comment (ralph-pichler) [inline swap/swap_test.go]: use `.Equal` here

File: swap/swap_test.go [hunk 4]
```diff
@@ -682,6 +688,125 @@ func TestResetBalance(t *testing.T) {
...
+
+	// set asymmetric balance and attempt to send cheque
+	if err = creditor.setBalance(-int64(ChequeDebtTolerance * 2)); err != nil {
+		t.Fatal(err)
+	}
+	// now simulate sending the cheque to the creditor from the debitor
+	if err = creditor.sendCheque(); err != nil {
+		t.Fatal(err)
+	}
+	cheque := creditor.getPendingCheque()
```
Comment (ralph-pichler) [inline swap/swap_test.go]: why do we need to create the cheque using `creditor`? wouldn't it be much simpler to just manually create the cheque (e.g. using `newSignedTestCheque`) and call `handleEmitChequeMsg` on `creditorSwap`? After all that's the only thing that really needs to be tested here.

File: swap/swap_test.go [hunk 5]
```diff
@@ -682,6 +688,125 @@ func TestResetBalance(t *testing.T) {
...
+	// simulate cheque handling
+	err = creditorSwap.handleEmitChequeMsg(ctx, debitor, &EmitChequeMsg{
+		Cheque: cheque,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// simulate cheque confirmation
+	debitorSwap.handleConfirmChequeMsg(ctx, creditor, &ConfirmChequeMsg{
```
Comment (ralph-pichler) [inline swap/swap_test.go]: this seems irrelevant to what this test is supposed to check.

**Review by ralph-pichler — CHANGES_REQUESTED** [review]
`TestDebtCheques` could be simplified significantly. Otherwise everything looks ok but please check the value of `ChequeDebtTolerance`.

File: swap/config.go [hunk 6]
```diff
@@ -27,6 +27,9 @@ const (
 	// DefaultPaymentThreshold is set to be equivalent to requesting and serving 10mb of data (2441 chunks (4096 bytes) = 10 mb, 10^7 bytes = 10 mb)
 	DefaultPaymentThreshold    = 2441*RetrieveRequestPrice + (10^7)*ChunkDeliveryPrice // 4096 * 2441 = 10 mb,
 	DefaultDisconnectThreshold = 20 * DefaultPaymentThreshold
+	// ChequeDebtTolerance is the lowest resulting balance a node is willing to accept when receiving a cheque
+	// the value is meant to be used below 0, as positive resulting balances should always be accepted when receiving cheques
+	ChequeDebtTolerance = DefaultPaymentThreshold / 20 // roughly 20% of the payment threshold
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/config.go]: embarrassing bug. 

fixed, thanks

File: swap/swap_test.go (hunk 2 above)
Comment (mortelli) [reply swap/swap_test.go]: done, thanks

File: swap/swap_test.go (hunk 3 above)
Comment (mortelli) [reply swap/swap_test.go]: done, thanks

File: swap/swap_test.go (hunk 5 above)
Comment (mortelli) [reply swap/swap_test.go]: if i remove this, i have to skip the balances verification. 

cheque confirmation will allow the creditor to reset its peer balance to 0.

i guess this isn't really what the test is about, so it can be skipped.

File: swap/swap_test.go (hunk 4 above)
Comment (mortelli) [reply swap/swap_test.go]: i honestly hadn't thought of this. i will reply down below.

**Comment by mortelli** [comment]
> `TestDebtCheques` could be simplified significantly. Otherwise everything looks ok but please check the value of `ChequeDebtTolerance`.

i have significantly simplified this function so that it is a unit test for debt cheques, and does not include other unrelated (and redundant) verifications. 

i have taken a look at the `newSignedTestCheque` function as you mention, but i don't feel it will particularly simplify the test as it currently stands. it also feels a bit out of place, given that no other test in `swap_test.go` uses this mechanism.

if you still think it's worth changing, i would like to discuss it further.

File: swap/swap_test.go (hunk 5 above)
Comment (mortelli) [reply swap/swap_test.go]: removed

File: swap/simulations_test.go (hunk 1 above)
Comment (mortelli) [reply swap/simulations_test.go]: removed the `Sleep` call and added a `allMessagesArrived` channel

File: swap/swap_test.go (hunk 5 above)
Comment (mortelli) [reply swap/swap_test.go]: see reply [here](https://github.com/ethersphere/swarm/pull/1983#issuecomment-576332679)

File: swap/swap_test.go (hunk 4 above)
Comment (mortelli) [reply swap/swap_test.go]: see reply [here](https://github.com/ethersphere/swarm/pull/1983#issuecomment-576332679)

File: swap/simulations_test.go [hunk 7]
```diff
@@ -491,6 +410,7 @@ func TestMultiChequeSimulation(t *testing.T) {
 	}
 
 	if b1 != -b2 {
+		fmt.Printf("balances")
```
Comment (ralph-pichler) [inline swap/simulations_test.go]: should this be here?

Comment (mortelli) [reply swap/simulations_test.go]: oh boy. removed, thanks

**Comment by mortelli** [comment]
created issue #2078 to follow-up on rare errors.

File: swap/swap_test.go (hunk 4 above)
Comment (ralph-pichler) [reply swap/swap_test.go]: the reason why it's not used anywhere else (yet) is because the helper function is still new (merged only last week I think). There are already many other tests that are more complicated than they need to be and we shouldn't add new ones.

Comment (mortelli) [reply swap/swap_test.go]: got it. i will simplify this test further by calling this function as soon as #2078 is closed

**Comment by mortelli** [comment]
@ralph-pichler i have pushed a simplified version of the `TestDebtCheques` function per your suggestion

**Review by ralph-pichler — APPROVED** [review]
LGTM

File: swap/simulations_test.go [hunk 8]
```diff
@@ -458,27 +330,74 @@ func TestMultiChequeSimulation(t *testing.T) {
...
-		if err := waitForChequeProcessed(t, params.backend, counter, lastCount, debitorSvc.swap.peers[creditor], expectedPayout); err != nil {
+
+		if err := creditorPeer.Send(context.Background(), &testMsgSmallPrice{}); err != nil {
 			t.Fatal(err)
 		}
-		lastCount += 1
-		expectedPayout += DefaultPaymentThreshold + 1
+
+		// check if cheque should have been sent
+		balanceAfterMessage := debitorBalance - int64(msgPrice)
```
Comment (holisticode) [inline swap/simulations_test.go:376]: Can we be sure at this point that the actual balance has been updated from the previous iteration run? I think it does, but give it another thought please.

File: swap/simulations_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/simulations_test.go]: hey, not sure how you came to view this version of the file, but this was already fixed [earlier](https://github.com/ethersphere/swarm/pull/1983#discussion_r368664374)

File: swap/simulations_test.go [hunk 9]
```diff
@@ -458,27 +330,74 @@ func TestMultiChequeSimulation(t *testing.T) {
...
+
+	for i := 0; i < msgAmount; i++ {
+		debitorBalance, err := debitorSvc.swap.loadBalance(creditor)
+		if err != nil {
 			t.Fatal(err)
 		}
-		// we need to wait a bit in order to give time for the cheque to be processed
-		if err := waitForChequeProcessed(t, params.backend, counter, lastCount, debitorSvc.swap.peers[creditor], expectedPayout); err != nil {
+
+		if err := creditorPeer.Send(context.Background(), &testMsgSmallPrice{}); err != nil {
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/simulations_test.go]: fixed

File: swap/simulations_test.go [hunk 10]
```diff
@@ -458,27 +330,74 @@ func TestMultiChequeSimulation(t *testing.T) {
...
+
+	minCheques := 2
+	maxCheques := 5
+
+	msgsPerCheque := (uint64(paymentThreshold) / msgPrice) + 1 // +1 to round up without casting to float
+
+	minMsgs := msgsPerCheque * uint64(minCheques)
+	maxMsgs := msgsPerCheque * uint64(maxCheques)
+
+	msgAmount := rand.Intn(int(maxMsgs-minMsgs)) + int(minMsgs)
```
[the remark this answers is not in the document]
Comment (mortelli) [reply swap/simulations_test.go]: it's true that i prefer fixed numbers... which doesn't really explain why i opted for randomness here.

changed

File: swap/simulations_test.go (hunk 8 above)
Comment (mortelli) [reply swap/simulations_test.go:376]: thank you for raising this point, it's something i hadn't considered.

my instinct was that the code (from this side of the exchange) would not be able to send a message in a following iteration until it was done with the current one, and this would include updating the balance.

i added some prints to the `updateBalance` calls and the test `for` loop iterations, and wrote an execution of `TestMultiChequeSimulation` test to disk.

[here](https://pastebin.com/raw/mGUqL7ak) are the initial results. you can see that sometimes there are multiple balance updates within a single iteration. however, these seem to be from the creditor side, so it does not matter that they occur one time per iteration—at least for the point you've raised. 

[this](https://pastebin.com/raw/DeAPCGr6) is the result obtained by filtering out the other peer's balance updates prints. it seems to confirm that every iteration does 1 balance update before moving on to the next one.

curiously, this might give us insight into the cause of #2078. it's very possible that by the time the creditor is allowed to receive the messages and account for them, the debitor is way ahead and the cheque it has sent is perceived as debt from the other side, for being too early. but i think it would have to process the cheque before doing the balance updates for this to explain the situation.

**Review by holisticode — APPROVED** [review]

**Comment by Eknir** [comment]
Something occurred to me:
Why would we disconnect from a node that sends us too much money? Surely, receiving too much money can only be good, right? I understand the context of this PR and how, when properly accounted for, receiving a cheque that is worth too much opens up the possibility for an attack. However, I believe we can solve this situation in a more elegant manner than a disconnect or putting the peer on a blacklist.

Note the sentence: 
>and how, when **properly accounted for**, receiving a cheque that is worth too much opens up the possibility for an attack.

I propose to do the following when we receive a cheque that is worth too much: don't account for it properly :).

Assume Node A has a balance of -500 with node B, node B sends a cheque of 10000 to A. If we would properly account for this, node A would immediately respond with a cheque to B as 9500 is surely more than the payment threshold. But, if we instead just program the node to update the balance to 100 instead of 9500, there is no attack possible anymore. The amount `100` can have the same value as what `ChequeDebtTolerance` has now.

If nodes are programmed to behave in this way, the attack is not possible anymore. In a normal situation, the situation above will not happen.

**Comment by mortelli** [comment]
> Something occurred to me:
> Why would we disconnect from a node that sends us too much money? Surely, receiving too much money can only be good, right? I understand the context of this PR and how, when properly accounted for, receiving a cheque that is worth too much opens up the possibility for an attack. However, I believe we can solve this situation in a more elegant manner than a disconnect or putting the peer on a blacklist.
> 
> Note the sentence:
> 
> > and how, when **properly accounted for**, receiving a cheque that is worth too much opens up the possibility for an attack.
> 
> I propose to do the following when we receive a cheque that is worth too much: don't account for it properly :).
> 
> Assume Node A has a balance of -500 with node B, node B sends a cheque of 10000 to A. If we would properly account for this, node A would immediately respond with a cheque to B as 9500 is surely more than the payment threshold. But, if we instead just program the node to update the balance to 100 instead of 9500, there is no attack possible anymore. The amount `100` can have the same value as what `ChequeDebtTolerance` has now.
> 
> If nodes are programmed to behave in this way, the attack is not possible anymore. In a normal situation, the situation above will not happen.

since we already have 2 approvals for this PR as-is, i suggest we open a different issue/PR for the solution you mention here.

i will merge the code in its current state as soon as i am able to, i.e. when `master` is fixed so that the CI passes.
